File: springboot/src/main/java/com/edufelip/meer/AppConfig.java
package com.edufelip.meer;

import com.edufelip.meer.domain.CreateCategoryUseCase;
import com.edufelip.meer.domain.CreateGuideContentCommentUseCase;
import com.edufelip.meer.domain.CreateGuideContentUseCase;
import com.edufelip.meer.domain.CreateThriftStoreUseCase;
import com.edufelip.meer.domain.DeleteCategoryUseCase;
import com.edufelip.meer.domain.GetCategoriesUseCase;
import com.edufelip.meer.domain.GetGuideContentUseCase;
import com.edufelip.meer.domain.GetGuideContentsByThriftStoreUseCase;
import com.edufelip.meer.domain.GetThriftStoreUseCase;
import com.edufelip.meer.domain.GetThriftStoresUseCase;
import com.edufelip.meer.domain.UpsertPushTokenUseCase;
import com.edufelip.meer.domain.UpdateCategoryUseCase;
import com.edufelip.meer.domain.UpdateGuideContentCommentUseCase;
import com.edufelip.meer.domain.DeletePushTokenUseCase;
import com.edufelip.meer.domain.auth.AppleLoginUseCase;
import com.edufelip.meer.domain.auth.DashboardLoginUseCase;
import com.edufelip.meer.domain.auth.ForgotPasswordUseCase;
import com.edufelip.meer.domain.auth.GetProfileUseCase;
import com.edufelip.meer.domain.auth.GoogleLoginUseCase;
import com.edufelip.meer.domain.auth.LoginUseCase;
import com.edufelip.meer.domain.auth.PasswordResetNotifier;
import com.edufelip.meer.domain.auth.RefreshTokenUseCase;
import com.edufelip.meer.domain.auth.ResetPasswordUseCase;
import com.edufelip.meer.domain.auth.SignupUseCase;
import com.edufelip.meer.domain.auth.UpdateProfileUseCase;
import com.edufelip.meer.domain.repo.AuthUserRepository;
import com.edufelip.meer.domain.repo.CategoryRepository;
import com.edufelip.meer.domain.repo.GuideContentCommentRepository;
import com.edufelip.meer.domain.repo.GuideContentRepository;
import com.edufelip.meer.domain.repo.PasswordResetTokenRepository;
import com.edufelip.meer.domain.repo.PushTokenRepository;
import com.edufelip.meer.domain.repo.ThriftStoreRepository;
import com.edufelip.meer.logging.RequestResponseLoggingFilter;
import com.edufelip.meer.config.FirebaseProperties;
import com.edufelip.meer.security.DashboardAdminGuardFilter;
import com.edufelip.meer.security.GoogleClientProperties;
import com.edufelip.meer.security.JwtProperties;
import com.edufelip.meer.security.PasswordResetProperties;
import com.edufelip.meer.security.RequestGuardsFilter;
import com.edufelip.meer.security.SecurityProperties;
import com.edufelip.meer.security.token.JwtTokenProvider;
import com.edufelip.meer.security.token.TokenProvider;
import java.time.Clock;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.boot.web.servlet.FilterRegistrationBean;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.Ordered;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;

@Configuration
@EnableCaching
@EnableConfigurationProperties({
  SecurityProperties.class,
  JwtProperties.class,
  GoogleClientProperties.class,
  PasswordResetProperties.class,
  FirebaseProperties.class
})
public class AppConfig {

  @Bean
  public GetThriftStoreUseCase getThriftStoreUseCase(ThriftStoreRepository repo) {
    return new GetThriftStoreUseCase(repo);
  }

  @Bean
  public GetThriftStoresUseCase getThriftStoresUseCase(
      ThriftStoreRepository repo,
      @org.springframework.beans.factory.annotation.Value("${spring.datasource.url:}")
          String datasourceUrl,
      @org.springframework.beans.factory.annotation.Value("${meer.postgis.enabled:false}")
          boolean postgisEnabled) {
    return new GetThriftStoresUseCase(repo, datasourceUrl, postgisEnabled);
  }

  @Bean
  public CreateThriftStoreUseCase createThriftStoreUseCase(ThriftStoreRepository repo) {
    return new CreateThriftStoreUseCase(repo);
  }

  @Bean
  public GetCategoriesUseCase getCategoriesUseCase(CategoryRepository repo) {
    return new GetCategoriesUseCase(repo);
  }

  @Bean
  public CreateCategoryUseCase createCategoryUseCase(CategoryRepository repo) {
    return new CreateCategoryUseCase(repo);
  }

  @Bean
  public UpdateCategoryUseCase updateCategoryUseCase(CategoryRepository repo) {
    return new UpdateCategoryUseCase(repo);
  }

  @Bean
  public DeleteCategoryUseCase deleteCategoryUseCase(CategoryRepository repo) {
    return new DeleteCategoryUseCase(repo);
  }

  @Bean
  public GetGuideContentUseCase getGuideContentUseCase(GuideContentRepository repo) {
    return new GetGuideContentUseCase(repo);
  }

  @Bean
  public GetGuideContentsByThriftStoreUseCase getGuideContentsByThriftStoreUseCase(
      GuideContentRepository repo) {
    return new GetGuideContentsByThriftStoreUseCase(repo);
  }

  @Bean
  public CreateGuideContentUseCase createGuideContentUseCase(
      GuideContentRepository repo, ThriftStoreRepository storeRepo) {
    return new CreateGuideContentUseCase(repo, storeRepo);
  }

  @Bean
  public CreateGuideContentCommentUseCase createGuideContentCommentUseCase(
      GuideContentCommentRepository repo,
      GuideContentRepository guideContentRepository,
      Clock clock) {
    return new CreateGuideContentCommentUseCase(repo, guideContentRepository, clock);
  }

  @Bean
  public UpdateGuideContentCommentUseCase updateGuideContentCommentUseCase(
      GuideContentCommentRepository repo, Clock clock) {
    return new UpdateGuideContentCommentUseCase(repo, clock);
  }

  @Bean
  public LoginUseCase loginUseCase(
      AuthUserRepository repo, PasswordEncoder encoder, TokenProvider tokenProvider) {
    return new LoginUseCase(repo, encoder, tokenProvider);
  }

  @Bean
  public DashboardLoginUseCase dashboardLoginUseCase(
      AuthUserRepository repo, PasswordEncoder encoder, TokenProvider tokenProvider) {
    return new DashboardLoginUseCase(repo, encoder, tokenProvider);
  }

  @Bean
  public SignupUseCase signupUseCase(
      AuthUserRepository repo, PasswordEncoder encoder, TokenProvider tokenProvider) {
    return new SignupUseCase(repo, encoder, tokenProvider);
  }

  @Bean
  public GoogleLoginUseCase googleLoginUseCase(
      AuthUserRepository repo,
      TokenProvider tokenProvider,
      PasswordEncoder encoder,
      GoogleClientProperties googleProps) {
    return new GoogleLoginUseCase(repo, tokenProvider, encoder, googleProps);
  }

  @Bean
  public AppleLoginUseCase appleLoginUseCase(
      AuthUserRepository repo, TokenProvider tokenProvider, PasswordEncoder encoder) {
    return new AppleLoginUseCase(repo, tokenProvider, encoder);
  }

  @Bean
  public RefreshTokenUseCase refreshTokenUseCase(
      TokenProvider tokenProvider, AuthUserRepository repo) {
    return new RefreshTokenUseCase(tokenProvider, repo);
  }

  @Bean
  public ForgotPasswordUseCase forgotPasswordUseCase(
      AuthUserRepository repo,
      PasswordResetTokenRepository passwordResetTokenRepository,
      PasswordResetNotifier passwordResetNotifier,
      PasswordResetProperties passwordResetProperties,
      Clock clock) {
    return new ForgotPasswordUseCase(
        repo, passwordResetTokenRepository, passwordResetNotifier, passwordResetProperties, clock);
  }

  @Bean
  public ResetPasswordUseCase resetPasswordUseCase(
      PasswordResetTokenRepository passwordResetTokenRepository,
      AuthUserRepository authUserRepository,
      PasswordEncoder passwordEncoder,
      Clock clock) {
    return new ResetPasswordUseCase(
        passwordResetTokenRepository, authUserRepository, passwordEncoder, clock);
  }

  @Bean
  public GetProfileUseCase getProfileUseCase(TokenProvider tokenProvider, AuthUserRepository repo) {
    return new GetProfileUseCase(tokenProvider, repo);
  }

  @Bean
  public UpdateProfileUseCase updateProfileUseCase(
      TokenProvider tokenProvider, AuthUserRepository repo) {
    return new UpdateProfileUseCase(tokenProvider, repo);
  }

  @Bean
  public UpsertPushTokenUseCase upsertPushTokenUseCase(
      PushTokenRepository pushTokenRepository, Clock clock) {
    return new UpsertPushTokenUseCase(pushTokenRepository, clock);
  }

  @Bean
  public DeletePushTokenUseCase deletePushTokenUseCase(PushTokenRepository pushTokenRepository) {
    return new DeletePushTokenUseCase(pushTokenRepository);
  }

  @Bean
  public PasswordEncoder passwordEncoder() {
    return new BCryptPasswordEncoder();
  }

  @Bean
  public Clock clock() {
    return Clock.systemUTC();
  }

  @Bean
  public TokenProvider tokenProvider(JwtProperties props, Clock clock) {
    return new JwtTokenProvider(props, clock);
  }

  @Bean
  public FilterRegistrationBean<RequestGuardsFilter> requestGuardsFilter(
      SecurityProperties securityProps,
      TokenProvider tokenProvider,
      AuthUserRepository authUserRepository) {
    FilterRegistrationBean<RequestGuardsFilter> registration =
        new FilterRegistrationBean<>(
            new RequestGuardsFilter(securityProps, tokenProvider, authUserRepository));
    registration.setOrder(Ordered.HIGHEST_PRECEDENCE);
    return registration;
  }

  @Bean
  public FilterRegistrationBean<RequestResponseLoggingFilter> requestResponseLoggingFilter() {
    FilterRegistrationBean<RequestResponseLoggingFilter> registration =
        new FilterRegistrationBean<>(new RequestResponseLoggingFilter());
    registration.setOrder(Ordered.HIGHEST_PRECEDENCE + 1); // run right after guards
    return registration;
  }

  @Bean
  public FilterRegistrationBean<DashboardAdminGuardFilter> dashboardAdminGuardFilter(
      TokenProvider tokenProvider, AuthUserRepository authUserRepository) {
    FilterRegistrationBean<DashboardAdminGuardFilter> registration =
        new FilterRegistrationBean<>(
            new DashboardAdminGuardFilter(tokenProvider, authUserRepository));
    registration.setOrder(Ordered.HIGHEST_PRECEDENCE + 2); // after logging
    return registration;
  }
}

File: springboot/src/main/kotlin/com/edufelip/meer/AppConfig.kt
package com.edufelip.meer

import com.edufelip.meer.domain.*
import com.edufelip.meer.security.GoogleClientProperties
import com.edufelip.meer.security.JwtProperties
import com.edufelip.meer.security.RequestGuardsFilter
import com.edufelip.meer.security.SecurityProperties
import com.edufelip.meer.security.token.JwtTokenProvider
import com.edufelip.meer.security.token.TokenProvider
import com.edufelip.meer.domain.auth.SignupUseCase
import com.edufelip.meer.domain.auth.GoogleLoginUseCase
import com.edufelip.meer.domain.auth.RefreshTokenUseCase
import com.edufelip.meer.domain.auth.AppleLoginUseCase
import com.edufelip.meer.domain.auth.ForgotPasswordUseCase
import com.edufelip.meer.domain.auth.LoginUseCase
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.boot.web.servlet.FilterRegistrationBean
import org.springframework.core.Ordered
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder

@Configuration
@EnableConfigurationProperties(value = [SecurityProperties::class, JwtProperties::class, GoogleClientProperties::class])
class AppConfig {

    @Bean
    fun getThriftStoreUseCase(thriftStoreRepository: ThriftStoreRepository): GetThriftStoreUseCase {
        return GetThriftStoreUseCase(thriftStoreRepository)
    }

    @Bean
    fun getThriftStoresUseCase(thriftStoreRepository: ThriftStoreRepository): GetThriftStoresUseCase {
        return GetThriftStoresUseCase(thriftStoreRepository)
    }

    @Bean
    fun createThriftStoreUseCase(thriftStoreRepository: ThriftStoreRepository): CreateThriftStoreUseCase {
        return CreateThriftStoreUseCase(thriftStoreRepository)
    }

    @Bean
    fun getGuideContentUseCase(guideContentRepository: GuideContentRepository): GetGuideContentUseCase {
        return GetGuideContentUseCase(guideContentRepository)
    }

    @Bean
    fun getGuideContentsByThriftStoreUseCase(guideContentRepository: GuideContentRepository): GetGuideContentsByThriftStoreUseCase {
        return GetGuideContentsByThriftStoreUseCase(guideContentRepository)
    }

    @Bean
    fun createGuideContentUseCase(
        guideContentRepository: GuideContentRepository,
        thriftStoreRepository: ThriftStoreRepository
    ): CreateGuideContentUseCase {
        return CreateGuideContentUseCase(guideContentRepository, thriftStoreRepository)
    }

    @Bean
    fun signupUseCase(
        authUserRepository: AuthUserRepository,
        passwordEncoder: PasswordEncoder,
        tokenProvider: TokenProvider
    ): SignupUseCase = SignupUseCase(authUserRepository, passwordEncoder, tokenProvider)

    @Bean
    fun googleLoginUseCase(
        authUserRepository: AuthUserRepository,
        tokenProvider: TokenProvider,
        passwordEncoder: PasswordEncoder,
        googleClientProperties: GoogleClientProperties
    ): GoogleLoginUseCase = GoogleLoginUseCase(authUserRepository, tokenProvider, passwordEncoder, googleClientProperties)

    @Bean
    fun refreshTokenUseCase(
        tokenProvider: TokenProvider,
        authUserRepository: AuthUserRepository
    ): RefreshTokenUseCase = RefreshTokenUseCase(tokenProvider, authUserRepository)

    @Bean
    fun appleLoginUseCase(
        authUserRepository: AuthUserRepository,
        tokenProvider: TokenProvider,
        passwordEncoder: PasswordEncoder
    ): AppleLoginUseCase = AppleLoginUseCase(authUserRepository, tokenProvider, passwordEncoder)

    @Bean
    fun forgotPasswordUseCase(
        authUserRepository: AuthUserRepository
    ): ForgotPasswordUseCase = ForgotPasswordUseCase(authUserRepository)

    @Bean
    fun passwordEncoder(): PasswordEncoder = BCryptPasswordEncoder()

    @Bean
    fun tokenProvider(jwtProperties: JwtProperties): TokenProvider = JwtTokenProvider(jwtProperties)

    @Bean
    fun loginUseCase(
        authUserRepository: AuthUserRepository,
        passwordEncoder: PasswordEncoder,
        tokenProvider: TokenProvider
    ): LoginUseCase = LoginUseCase(authUserRepository, passwordEncoder, tokenProvider)

    @Bean
    fun requestGuardsFilter(securityProperties: SecurityProperties): FilterRegistrationBean<RequestGuardsFilter> {
        val registration = FilterRegistrationBean(RequestGuardsFilter(securityProperties))
        registration.order = Ordered.HIGHEST_PRECEDENCE
        return registration
    }
}
